File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    from collections import deque

    q=deque()
    q.append((begin,0))
    visited = set()
    answer = 0
    
    if target not in words:
        return 0
    
    def check(a,b):
        diff = 0
        for i in range(len(a)):
            if a[i] != b[i]:
                diff += 1
                if diff > 1:
                    return False

        return diff == 1
    
    while q:
        now, count = q.popleft()

        if now == target:
            answer = count

        for word in words:
            if word not in visited and check(now,word):
                visited.add(word)
                q.append((word,count+1))
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>

using namespace std;

bool visited[50];
int answer = 50;

bool check(string A, string B){
    int cnt = 0;
    for(int i=0; i<A.size();i++){
        if(A[i]!=B[i]) cnt++;
        if(cnt>1) return false;
    }
    return true;
}


void dfs(string begin, string target, vector<string>& words, int level){
    if(answer <= level) return;
    if(begin == target){
        answer = min(answer, level);
        return;
    }
    
    for(int i=0;i<words.size();i++){
        if(check(begin, words[i]) && !visited[i]){
            visited[i] = true;
            dfs(words[i],target,words,level+1);
            visited[i] = false;
        }
    }
}


int solution(string begin, string target, vector<string> words) {
    dfs(begin, target, words, 0);
    if(answer==50) answer = 0;
    
    return answer;
}